rust/arrow/src/compute/kernels/filter.rs

```diff
@@ -17,841 +17,241 @@
-    /// selected by the BooleanArray passed as the filter_array parameter
-    pub fn filter_primitive_array<T>(
-        &self,
-        data_array: &PrimitiveArray<T>,
-    ) -> Result<PrimitiveArray<T>>
-    where
-        T: ArrowNumericType,
-    {
-        let array_type = T::DATA_TYPE;
-        let value_size = mem::size_of::<T::Native>();
-        let array_data_builder =
-            filter_array_impl(self, data_array, array_type, value_size)?;
-        let data = array_data_builder.build();
-        Ok(PrimitiveArray::<T>::from(data))
-    }
-
-    /// Returns a new DictionaryArray<T> containing only those keys from the array passed as the data_array parameter,
-    /// selected by the BooleanArray passed as the filter_array parameter. The values are cloned from the data_array.
-    pub fn filter_dictionary_array<T>(
-        &self,
-        data_array: &DictionaryArray<T>,
-    ) -> Result<DictionaryArray<T>>
-    where
-        T: ArrowNumericType,
-    {
-        let array_type = data_array.data_type().clone();
-        let value_size = mem::size_of::<T::Native>();
-        let mut array_data_builder =
-            filter_array_impl(self, data_array, array_type, value_size)?;
-        // copy dictionary values from input array
-        array_data_builder =
-            array_data_builder.add_child_data(data_array.values().data());
-        let data = array_data_builder.build();
-        Ok(DictionaryArray::<T>::from(data))
     }
 }
 
-/// Returns a new array, containing only the elements matching the filter.
-pub fn filter(array: &Array, filter: &BooleanArray) -> Result<ArrayRef> {
-    FilterContext::new(filter)?.filter(array)
+/// Returns a function used to filter arbitrary arrays.
+/// This is faster (2x for primitive types) than using [filter] on multiple arrays, but slower
+/// than [filter] when filtering a single array.
+pub fn build_filter(filter: &BooleanArray) -> Result<Filter> {
+    let iter = SlicesIterator::new(filter);
+    let filter_count = iter.filter_count;
+    let chunks = iter.collect::<Vec<_>>();
+
+    Ok(Box::new(move |array: &ArrayData| {
+        let mut mutable = MutableArrayData::new(vec![array], false, filter_count);
+        chunks
+            .iter()
+            .for_each(|(start, end)| mutable.extend(0, *start, *end));
+        mutable.freeze()
+    }))
 }
 
-/// Returns a new PrimitiveArray<T> containing only those values from the array passed as the data_array parameter,
-/// selected by the BooleanArray passed as the filter_array parameter
-pub fn filter_primitive_array<T>(
-    data_array: &PrimitiveArray<T>,
-    filter_array: &BooleanArray,
-) -> Result<PrimitiveArray<T>>
-where
-    T: ArrowNumericType,
-{
-    FilterContext::new(filter_array)?.filter_primitive_array(data_array)
-}
+/// Filters an [Array], returning elements matching the filter (i.e. where the values are true).
+/// WARNING: the nulls of `filter` are ignored and the value on its slot is considered.
```

Comment:
I think this WARNING should also be included in the doc comments of `build_filter` as well

Reply:
Or instead of ignoring the nulls, we could perform a `values_buffer & null_buffer` if there are nulls. Given that the input `filter` will almost always be the result of some computation, I'd prefer that we incur the slight cost of the `AND` operation, so that we treat null slots as `false`.

Reply:
This "WARNING" was more like a way of describing a (unknown to me) feature. master already does this, but I did not have the time to go about checking what other implementations do (whether the result is null or something else).

Note that this hits datafusion already: there is at least one test there where we ignore a null value from a filter (because the predicate is built using arrays with nulls).